### PR TITLE
WIP Basic changefeed implementation 

### PR DESCRIPTION
**Pull request by gapra-msft**
Test situation still being worked out

This PR covers - basic CF from start to end, user settable start and end time, basic paging functionality.

### CODE REVIEW

File: sdk/storage/azure-storage-blob-changefeed/pom.xml [hunk 1]
```diff
@@ -94,6 +94,12 @@
       <version>1.0.4</version> <!-- {x-version-update;com.azure:azure-identity;dependency} -->
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/pom.xml:97]: This will eventually be removed once I implement my avro parser

File: sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/APISpec.groovy [hunk 2]
```diff
@@ -3,12 +3,236 @@
...
+        return getServiceClientBuilder(credential, String.format(defaultEndpointTemplate, credential.getAccountName()))
+            .buildAsyncClient()
+    }
+
+    BlobServiceClientBuilder getServiceClientBuilder(StorageSharedKeyCredential credential, String endpoint,
+                                                     HttpPipelinePolicy... policies) {
+        BlobServiceClientBuilder builder = new BlobServiceClientBuilder()
+            .endpoint(endpoint)
+            .httpClient(getHttpClient())
+        .serviceVersion(BlobServiceVersion.V2019_02_02)
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/APISpec.groovy:210]: undo this change

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    /**
+     * Validates that changefeed has been enabled for the account.
+     */
+    private Mono<Void> validateChangefeed() {
+        return this.client.exists()
+            .flatMap(exists -> {
+                if (exists == null || !exists) {
+                    throw logger.logExceptionAsError(new RuntimeException("ChangeFeed has not been enabled for this "
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java:95]: I think I should just be returning a Mono.error here

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/implementation/util/TimeUtils.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+     * <p>OR
+     * <p>Year path : idx/segments/1601/
+     * <p>OffsetDateTime : year - 1601, month - 00, day - 00, hour - 00, minute - 00
+     *
+     * @param path The path to convert.
+     * @return The time associated with the path.
+     */
+    public static OffsetDateTime convertPathToTime(String path) {
+        if (path == null) {
+            return null;
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/implementation/util/TimeUtils.java:23]: should I actually throw for the null inputs?

File: sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/BlobChangefeedCursorTest.groovy [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.azure.storage.blob.changefeed
+
+
+import com.azure.storage.blob.changefeed.implementation.util.BlobChangefeedCursor
+
+import java.time.OffsetDateTime
+
+class BlobChangefeedCursorTest extends APISpec {
+
+    def "test serialize"() {
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/BlobChangefeedCursorTest.groovy:10]: ignore this : need to formalize testing

File: sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/BlobChangefeedTest.groovy [hunk 6]
```diff
@@ -0,0 +1,446 @@
+package com.azure.storage.blob.changefeed
+
+class BlobChangefeedTest extends APISpec {
+
+//    def "basic sync implementation"() {
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/test/java/com/azure/storage/blob/changefeed/BlobChangefeedTest.groovy:5]: ignore this : need to formalize testing

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+        /* TODO (gapra) : Continuation token is currently ignored. */
+//        if (continuationToken != null) {
+//            this.changefeed = new Changefeed(client, continuationToken);
+//        }
+        if (preferredPageSize <= 0) {
+            return Flux.error(new IllegalArgumentException("preferredPageSize > 0 required but provided: "
+                + preferredPageSize));
+        }
+        preferredPageSize = Integer.min(preferredPageSize, defaultPageSize);
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java:75]: Why don't we just take the preferredPageSize?

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        return changefeed.getEvents()
+            /* Window the events to the page size. */
+            .window(preferredPageSize)
+            /* Convert the BlobChangefeedEventWrappers into BlobChangefeedEvents along with the end cursor. */
+            .flatMap(eventWrappers -> {
+                Flux<BlobChangefeedEventWrapper> c1 = eventWrappers.cache();
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java:82]: I think it's generally helpful to leave some more thorough comments on reactive sequences like this that describe what it does, memory/concurrency behavior, possible special cases and how we addressed them. It just helps us later when we have to update the code so we don't have to try to understand the nuance of all the operators all over again.

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedResponse.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+public class BlobChangefeedPagedResponse implements ContinuablePage<String, BlobChangefeedEvent> {
+
+    private final List<BlobChangefeedEvent> events;
+    private final BlobChangefeedCursor cursor;
+
+    BlobChangefeedPagedResponse(List<BlobChangefeedEvent> events, BlobChangefeedCursor cursor) {
+        this.events = events;
+        this.cursor = cursor;
+        System.out.println(events.get(0).getId());
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedResponse.java:18]: Remove

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    private final Integer defaultPageSize = 5000;
+
+    /**
+     * Creates an instance of {@link BlobChangefeedPagedFlux}.
+     * @param client The {@link BlobContainerAsyncClient changefeed client}.
+     * @param startTime The {@link OffsetDateTime start time}.
+     * @param endTime The {@link OffsetDateTime end time}.
+     */
+    BlobChangefeedPagedFlux(BlobContainerAsyncClient client, OffsetDateTime startTime, OffsetDateTime endTime) {
+//        this.client = client;
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChangefeedPagedFlux.java:37]: Remove

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Segment.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * Get all the events for the Segment.
+     * @return A reactive stream of {@link BlobChangefeedEventWrapper}
+     */
+    Flux<BlobChangefeedEventWrapper> getEvents() {
+        /* Download JSON manifest file. */
+        return client.getBlobAsyncClient(path)
+            .download().reduce(new ByteArrayOutputStream(), (os, buffer) -> {
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Segment.java]: Seems like it may be worth throwing this download->reduce->map stuff in a shared method

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Segment.java]: I think I already did that for all the ones that download a JSON file. Maybe refresh the page?
